README.md
# pytorch-nncp


Changelog
2021-02-06:

- added a larger model (nncp_enwik_large.sh)

2021-01-03:

- initial release


cutils.h
#ifndef CUTILS_H
#define CUTILS_H

#define force_inline inline __attribute__((always_inline))
#define no_inline __attribute__((noinline))
#define __unused __attribute__((unused))
#define xglue(x, y) x ## y
#define glue(x, y) xglue(x, y)
#ifndef offsetof
#define offsetof(type, field) ((size_t) &((type *)0)->field)
#endif
#define countof(x) (sizeof(x) / sizeof(x[0]))
#define likely(x) __builtin_expect(!!(x), 1)
#define unlikely(x) __builtin_expect(!!(x), 0)

typedef int BOOL;

#ifndef FALSE
enum {
    FALSE = 0,
    TRUE = 1,
};
#endif

#if defined(__x86_64__)
static inline int64_t get_cycles(void)
{
    uint32_t low,high;
    int64_t val;
    asm volatile("rdtsc" : "=a" (low), "=d" (high));
    val = high;
    val <<= 32;
    val |= low;
    return val;
}
#else
static inline int64_t get_cycles(void)
{
    int64_t val;
    asm volatile ("rdtsc" : "=A" (val));
    return val;
}
#endif

static inline int max_int(int a, int b)
{
    if (a > b)
        return a;
    else
        return b;
}

static inline int min_int(int a, int b)
{
    if (a < b)
        return a;
    else
        return b;
}

static inline size_t max_size_t(size_t a, size_t b)
{
    if (a > b)
        return a;
    else
        return b;
}

static inline size_t min_size_t(size_t a, size_t b)
{
    if (a < b)
        return a;
    else
        return b;
}

static inline ssize_t max_ssize_t(ssize_t a, ssize_t b)
{
    if (a > b)
        return a;
    else
        return b;
}

static inline ssize_t min_ssize_t(ssize_t a, ssize_t b)
{
    if (a < b)
        return a;
    else
        return b;
}

static inline int clamp_int(int val, int min_val, int max_val)
{
    if (val < min_val)
        return min_val;
    else if (val > max_val)
        return max_val;
    else
        return val;
}

static inline float clamp_float(float val, float min_val, float max_val)
{
    if (val < min_val)
        return min_val;
    else if (val > max_val)
        return max_val;
    else
        return val;
}

static inline float squaref(float x)
{
    return x * x;
}

#define DUP8(a) a, a, a, a, a, a, a, a

#endif /* CUTILS_H */



exp_utils.py
# from https://github.com/kimiyoung/transformer-xl
import functools
import os, shutil

import numpy as np

import torch


def logging(s, log_path, print_=True, log_=True):
    if print_:
        print(s)
    if log_:
        with open(log_path, 'a+') as f_log:
            f_log.write(s + '\n')

def get_logger(log_path, **kwargs):
    return functools.partial(logging, log_path=log_path, **kwargs)

def create_exp_dir(dir_path, scripts_to_save=None, debug=False):
    if debug:
        print('Debug Mode : no experiment dir created')
        return functools.partial(logging, log_path=None, log_=False)

    if not os.path.exists(dir_path):
        os.makedirs(dir_path)

    print('Experiment dir : {}'.format(dir_path))
    if scripts_to_save is not None:
        script_path = os.path.join(dir_path, 'scripts')
        if not os.path.exists(script_path):
            os.makedirs(script_path)
        for script in scripts_to_save:
            dst_file = os.path.join(dir_path, 'scripts', os.path.basename(script))
            shutil.copyfile(script, dst_file)

    return get_logger(log_path=os.path.join(dir_path, 'log.txt'))

def save_checkpoint(model, optimizer, path, epoch):
    torch.save(model, os.path.join(path, 'model_{}.pt'.format(epoch)))
    torch.save(optimizer.state_dict(), os.path.join(path, 'optimizer_{}.pt'.format(epoch)))


nncp_enwik_base.sh
#!/bin/bash
#
# NNCP v2 enwik8/enwik9 compression and decompression
#
cmd="$1"
file="$2"

if [ '(' "$cmd" != "c" -a "$cmd" != "d" ')' -o \
     '(' "$file" != "enwik8" -a "$file" != "enwik9" ')' ]; then
    echo "usage: nncp_enwik.sh [c|d] [enwik8|enwik9]"
    echo "NNCP v2 enwik8/enwik9 compression test"
    echo "c   compress 'file' to 'file.bin' and 'file.voc'"
    echo "d   decompress 'file.bin' and 'file.voc' to 'file.out'"
    echo "    with file=enwik8 or enwik9"
    exit 1
fi

infile=${file}
vocfile=${file}.voc
ppfile=${file}.pp
cmpfile=${file}.bin
ppoutfile=${file}.ppout
outfile=${file}.out

if [ "$cmd" == "c" ]; then
  if [ "$file" == "enwik8" ]; then
      ./preprocess c $vocfile $infile $ppfile 16384 64
  else
      ./preprocess c $vocfile $infile $ppfile 16384 512
  fi
fi

if [ "$cmd" == "c" ]; then
    nncp_args="--input $ppfile --output $cmpfile"
else
    nncp_args="--decompress --input $cmpfile --output $ppoutfile"
fi

# needed for deterministic PyTorch
export CUBLAS_WORKSPACE_CONFIG=:4096:8

python nncp.py $nncp_args \
       --debug \
       --cuda \
       --fp16 \
       --vocab_size 16388  \
       --block_len 500000 \
       --n_layer 12 \
       --d_model 512 \
       --n_head 8 \
       --d_head 64 \
       --d_inner 2048 \
       --attn_type 1 \
       --tied_r_bias \
       --dropout 0.25 \
       --dropatt 0.0 \
       --lr 7.9e-5,341105,1.6e-5,3134681,5.0e-6 \
       --tgt_len 1 \
       --ext_tgt_len 31 \
       --mem_len 160 \
       --batch_size 64 \
       --init_std 0.013 \
       --adam_eps 1e-9 \
       --clip 0.25 \
       --retrain_period 500000 \
       --retrain_len 10000000 \
       --retrain_tgt_len 64 \
       --retrain_mem_len 128 \
       --retrain_batch_size 32 \
       --retrain_lr 4.0e-4,13000,2.0e-4,93000,1.0e-4,163000,5.0e-5,1911300,1.6e-5 \
       --gelu


if [ "$cmd" == "d" ]; then
  ./preprocess d $vocfile $ppoutfile $outfile
fi


nncp_enwik_large.sh
#!/bin/bash
#
# NNCP v2 enwik8/enwik9 compression and decompression
#
cmd="$1"
file="$2"

if [ '(' "$cmd" != "c" -a "$cmd" != "d" ')' -o \
     '(' "$file" != "enwik8" -a "$file" != "enwik9" ')' ]; then
    echo "usage: nncp_enwik.sh [c|d] [enwik8|enwik9]"
    echo "NNCP v2 enwik8/enwik9 compression test"
    echo "c   compress 'file' to 'file.bin' and 'file.voc'"
    echo "d   decompress 'file.bin' and 'file.voc' to 'file.out'"
    echo "    with file=enwik8 or enwik9"
    exit 1
fi

infile=${file}
vocfile=${file}.voc
ppfile=${file}.pp
cmpfile=${file}.bin
ppoutfile=${file}.ppout
outfile=${file}.out

if [ "$cmd" == "c" ]; then
  if [ "$file" == "enwik8" ]; then
      ./preprocess c $vocfile $infile $ppfile 4096 512
  else
      ./preprocess c $vocfile $infile $ppfile 16384 512
  fi
fi

if [ "$cmd" == "c" ]; then
    nncp_args="--input $ppfile --output $cmpfile"
else
    nncp_args="--decompress --input $cmpfile --output $ppoutfile"
fi

if [ "$file" == "enwik8" ]; then
    vocab_size="4096"
    lr="4.0e-5,341105,1.3e-5,3134681,4.0e-6"
    retrain_len="15000000"
else
    vocab_size="16388"
    lr="4.0e-5,341105,1.3e-5,707961,7.9e-6,3134681,4.0e-6"
    retrain_len="5000000"
fi


# needed for deterministic PyTorch
export CUBLAS_WORKSPACE_CONFIG=:4096:8

python nncp.py $nncp_args \
       --debug \
       --cuda \
       --fp16 \
       --vocab_size $vocab_size  \
       --block_len 500000 \
       --n_layer 12 \
       --d_model 1024 \
       --n_head 8 \
       --d_head 128 \
       --d_inner 4096 \
       --attn_type 1 \
       --tied_r_bias \
       --dropout 0.25 \
       --dropatt 0.0 \
       --lr $lr \
       --tgt_len 1 \
       --ext_tgt_len 31 \
       --mem_len 160 \
       --batch_size 64 \
       --init_std 0.013 \
       --adam_eps 1e-9 \
       --clip 0.25 \
       --retrain_period 500000 \
       --retrain_len $retrain_len \
       --retrain_tgt_len 64 \
       --retrain_mem_len 128 \
       --retrain_batch_size 32 \
       --retrain_lr 1.6e-4,13000,1.6e-4,93000,7.9e-5,163000,4.0e-5,1911300,1.3e-5 \
       --gelu


if [ "$cmd" == "d" ]; then
  ./preprocess d $vocfile $ppoutfile $outfile
fi
